query-algebrizer/src/clauses/where_fn.rs

```diff
@@ -0,0 +1,486 @@
+                                                }
+                                            }
+                                            r
+                                        })
+                                        .collect::<Result<Vec<_>>>()?;
+
+                // Otherwise, we now have the values and the type.
+                let types = vec![accumulated_types.exemplar().unwrap()];
+                let names = vec![var.clone()];
+
+                self.collect_named_bindings(schema, names, types, values);
+                Ok(())
+            },
+
+            (Binding::BindRel(places), FnArg::Vector(rows)) => {
+                if rows.is_empty() {
+                    bail!(ErrorKind::InvalidGroundConstant);
+                }
+
+                // Grab the known types to which these args must conform, and track
+                // the places that won't be bound in the output.
+                let template: Vec<(bool, ValueTypeSet)> =
+                    places.iter()
+                          .map(|x| match x {
+                              &VariableOrPlaceholder::Placeholder     => (false, ValueTypeSet::any()),
+                              &VariableOrPlaceholder::Variable(ref v) => (true, self.known_type_set(v)),
+                          })
+                          .collect();
+
+                // The expected 'width' of the matrix is the number of named variables.
+                let full_width = places.len();
+                let names: Vec<Variable> = places.into_iter().filter_map(|x| x.into_var()).collect();
+                let expected_width = names.len();
+                let expected_rows = rows.len();
+
+                if expected_width == 0 {
+                    // They can't all be placeholders.
+                    bail!(ErrorKind::InvalidGroundConstant);
+                }
+
+                // Accumulate values into `matrix` and types into `a_t_f_c`.
+                // This representation of a rectangular matrix is more efficient than one composed
+                // of N separate vectors.
+                let mut matrix = Vec::with_capacity(expected_width * expected_rows);
+                let mut accumulated_types_for_columns: Vec<ValueTypeSet> =
+                    ::std::iter::repeat(ValueTypeSet::default())
+                        .take(expected_width)
+                        .collect();
+
+                // Loop so we can bail out.
+                for row in rows.into_iter() {
+                    match row {
+                        FnArg::Vector(cols) => {
+                            // Make sure that every row is the same length.
+                            if cols.len() != full_width {
+                                bail!(ErrorKind::InvalidGroundConstant);
+                            }
+
+                            // Convert each item in the row.
+                            let r: Result<Vec<TypedValue>>;
+                            r = cols.into_iter()
```

Comment:
This expression is awkward.  I think you'd do better to use the turbofish and `?`directly here, rather than staging it.